src/audio/chain_dma.c

```diff
@@ -0,0 +1,596 @@
+	memset(dma_block_cfg_link, 0, sizeof(*dma_block_cfg_link));
+	dma_cfg_link->block_count = 1;
+	dma_cfg_link->source_data_size = cd->cs;
+	dma_cfg_link->dest_data_size = cd->cs;
+	dma_cfg_link->head_block  = dma_block_cfg_link;
+	dma_block_cfg_link->block_size = length;
+
+	switch (cd->stream_direction) {
+	case SOF_IPC_STREAM_PLAYBACK:
+		dma_cfg_host->channel_direction = HOST_TO_MEMORY;
+		dma_block_cfg_host->dest_address = (uint32_t)addr;
+		dma_cfg_link->channel_direction = MEMORY_TO_PERIPHERAL;
+		dma_block_cfg_link->source_address = (uint32_t)addr;
+		break;
+	case SOF_IPC_STREAM_CAPTURE:
+		dma_cfg_host->channel_direction = MEMORY_TO_HOST;
+		dma_block_cfg_host->source_address = (uint32_t)addr;
+		dma_cfg_link->channel_direction = PERIPHERAL_TO_MEMORY;
+		dma_block_cfg_link->dest_address = (uint32_t)addr;
+		break;
+	}
+
+	/* get host DMA channel */
+	channel = cd->host_connector_node_id.f.v_index;
+	channel = dma_request_channel(cd->dma_host->z_dev,
+				      &channel);
+	if (channel < 0) {
+		comp_err(dev, "chain_init(): dma_request_channel() failed");
+		return -EINVAL;
+	}
+
+	cd->chan_host = &cd->dma_host->chan[channel];
+
+	err = dma_config(cd->dma_host->z_dev, cd->chan_host->index, dma_cfg_host);
+	if (err < 0) {
+		comp_err(dev, "chain_init(): dma_config() failed");
+		goto error_host;
+	}
+
+	/* get link DMA channel */
+	channel = cd->link_connector_node_id.f.v_index;
+	channel = dma_request_channel(cd->dma_link->z_dev,
+				      &channel);
+	if (channel < 0) {
+		comp_err(dev, "chain_init(): dma_request_channel() failed");
+		goto error_host;
+	}
+
+	cd->chan_link = &cd->dma_link->chan[channel];
+
+	err = dma_config(cd->dma_link->z_dev, cd->chan_link->index, dma_cfg_link);
+	if (err < 0) {
+		comp_err(dev, "chain_init(): dma_config() failed");
+		dma_release_channel(cd->dma_link->z_dev, cd->chan_link->index);
+		goto error_link;
+	}
+	return 0;
+
+error_link:
+	dma_release_channel(cd->dma_link->z_dev, cd->chan_link->index);
+	cd->chan_link = NULL;
+error_host:
+	dma_release_channel(cd->dma_host->z_dev, cd->chan_host->index);
+	cd->chan_host = NULL;
+	return -EINVAL;
+}
+
+size_t chain_get_transferred_data_size(const uint32_t out_read_pos, const uint32_t in_read_pos,
+				       const size_t buff_size)
+{
+	if (out_read_pos >= in_read_pos)
+		return out_read_pos - in_read_pos;
+
+	return (buff_size - in_read_pos) + out_read_pos;
+}
+
+int get_connector_node_id(uint32_t dma_id, bool host_type,
+			  union ipc4_connector_node_id *connector_node_id)
+{
+	uint8_t type = host_type ? ipc4_hda_host_output_class : ipc4_hda_link_output_class;
+
+	if (dma_id >= DAI_NUM_HDA_OUT) {
+		type = host_type ? ipc4_hda_host_input_class : ipc4_hda_link_input_class;
+		dma_id = dma_id - DAI_NUM_HDA_OUT;
+		if (dma_id >= DAI_NUM_HDA_IN)
+			return -EINVAL;
+	}
+	connector_node_id->dw = 0;
+	connector_node_id->f.dma_type = type;
+	connector_node_id->f.v_index = dma_id;
+
+	return 0;
+}
+
+static const struct comp_driver comp_chain_dma = {
```

Comment:
@makarukp the comp_driver model/API is being deprecated in favor of the module API, i.e. these 3 ops will need changed to be module API ops either now or later.

Reply:
I see... Do you mean by changing it to module_interface from module_interface.h like in mixin_mixout module ?

Reply:
I decided that due to lack of extended module api (extra dependency under discussion), transfer to module API will be added in another PR.